### PR TITLE
feat: Add generic types for NhostFunctionsClient and NhostGraphqlClient request method return type and variables

### PR DESCRIPTION
**Pull request by hajek-raven**
Add generic types for NhostClient request return type and variables. This improves developer experience.

Fixes https://github.com/nhost/nhost/issues/404

(My first PR in this project 🎉 )

### CODE REVIEW

**Review by szilarddoro — COMMENTED** [review]
Your PR is greatly appreciated @hajek-raven! Congrats on your **first one** for Nhost 🎉

Let me know what you think about my suggestions.

File: packages/nhost-js/src/clients/graphql.ts [hunk 1]
```diff
@@ -66,7 +66,7 @@ export class NhostGraphqlClient {
       }
     }
 
-    responseData = responseData as GraphqlResponse
+    responseData = responseData as GraphqlResponse<T>
```
[the remark this answers is not in the document]
Comment (hajek-raven) [reply packages/nhost-js/src/clients/graphql.ts]: Thank you for your feedback!

I decided to refactor the code as it is was chaotic. I got rid of the responseData in outer context altogether.

[the remark this answers is not in the document]
Comment (hajek-raven) [reply packages/nhost-js/src/clients/graphql.ts]: I think I'd be a good idea to refactor this class a bit further (improve testability).
One opportunity might be to add the ability to provide custom axios instance. Or maybe abstract the network library altogether, but that should be handeled in different PR. 😄

File: packages/nhost-js/src/clients/graphql.ts [hunk 2]
```diff
@@ -21,33 +21,47 @@ export class NhostGraphqlClient {
...
-      const res = await this.instance.post(
+      const operationName = ''
+      const res = await this.instance.post<GraphqlResponse<T>>(
         '',
         {
           operationName: operationName || undefined,
           query: document,
-          // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
-          variables
+          ...(variables ? { variables } : {})
```
Comment (szilarddoro) [inline packages/nhost-js/src/clients/graphql.ts:42]: For the sake of readability we can rather use `variables: variables || {}` here. I don't worry about having a `variables: {}` passed to the Axios POST call.

File: packages/nhost-js/src/clients/functions.ts [hunk 3]
```diff
@@ -18,7 +18,7 @@ export class NhostFunctionsClient {
     })
   }
 
-  async call(url: string, data: any, config?: AxiosRequestConfig): Promise<FunctionCallResponse> {
+  async call<T = unknown, D = any>(url: string, data: D, config?: AxiosRequestConfig): Promise<FunctionCallResponse<T>> {
```
Comment (szilarddoro) [inline packages/nhost-js/src/clients/functions.ts:21]: You did a good job on refactoring the `request` function in `graphql.ts`. 👍🏼 

Could you please apply the same changes to this function too? Then I think we are good to go.

**Review by szilarddoro — CHANGES_REQUESTED** [review]

File: packages/nhost-js/src/clients/graphql.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (szilarddoro) [reply packages/nhost-js/src/clients/graphql.ts]: Definitely agree 👍🏼 Don't hesitate to open an issue with your enhancement proposals. We can discuss the matter there.

**Review by szilarddoro — APPROVED** [review]

**Comment by szilarddoro** [comment]
Awesome 👍🏼  Thank you for your help @hajek-raven!

**Pull request merged**